gson/src/main/java/com/google/gson/internal/reflect/ReflectionHelper.java

```diff
@@ -65,10 +82,131 @@ public static String tryMakeAccessible(Constructor<?> constructor) {
+      IllegalAccessException exception) {
     throw new RuntimeException("Unexpected IllegalAccessException occurred (Gson " + GsonBuildConfig.VERSION + "). "
         + "Certain ReflectionAccessFilter features require Java >= 9 to work correctly. If you are not using "
         + "ReflectionAccessFilter, report this to the Gson maintainers.",
         exception);
   }
+
+  public static <T> Constructor<T> getCanonicalRecordConstructor(Class<T> raw) {
+    return RECORD_HELPER.getCanonicalRecordConstructor(raw);
+  }
+
+  /**
+   * Internal abstraction over reflection when Records are supported.
+   */
+  private abstract static class RecordHelper {
+    abstract boolean isRecord(Class<?> clazz);
+
+    abstract String[] recordFields(Class<?> clazz);
+
+    abstract <T> Constructor<T> getCanonicalRecordConstructor(Class<T> raw);
+  }
+
+  private static class RecordSupportedHelper extends RecordHelper {
+    private final Method isRecord;
+    private final Method getRecordComponents;
+    private final Method getName;
+    private final Method getType;
+
+    private RecordSupportedHelper() throws NoSuchMethodException {
+      isRecord = Class.class.getMethod("isRecord");
+      getRecordComponents = Class.class.getDeclaredMethod("getRecordComponents");
+      Class<?> recordComponentType = getRecordComponents.getReturnType().getComponentType();
+      getName = recordComponentType.getDeclaredMethod("getName");
+      getType = recordComponentType.getDeclaredMethod("getType");
+    }
+
+    @Override
+    boolean isRecord(Class<?> raw) {
+      try {
+        return boolean.class.cast(isRecord.invoke(raw));
+      } catch (IllegalAccessException e) {
+        throw createExceptionForUnexpectedIllegalAccess(e);
+      } catch (InvocationTargetException e) {
+        throw new RuntimeException("Failed to reflect into record components on"
+                + " class [" + raw + "], this is unexpected behaviour, as these methods should"
+                + " not throw when they are defined on Class",
+                e
+        );
+      }
+    }
+
+    @Override
+    String[] recordFields(Class<?> raw) {
+      try {
+        Object recordComponents = getRecordComponents.invoke(raw);
+        int componentCount = Array.getLength(recordComponents);
+        String[] recordFields = new String[componentCount];
+        for (int i = 0; i < componentCount; i++) {
+          recordFields[i] =
+                  String.class.cast(getName.invoke(Array.get(recordComponents, i)));
+        }
+        return recordFields;
+      } catch (IllegalAccessException e) {
+        throw createExceptionForUnexpectedIllegalAccess(e);
+      } catch (ReflectiveOperationException e) {
+        throw new RuntimeException("Failed to reflect into record components on"
+                + " class [" + raw + "], this is unexpected behaviour, as these methods should"
+                + " not throw when they are defined on Class",
```

Comment:
If you want you can include `(Gson " + GsonBuildConfig.VERSION + ")` in the message; maybe that will help troubleshooting in the future.